font/util/man/bdftruncate.man
.TH BDFTRUNCATE __appmansuffix__ 2025-04-27 __vendorversion__
.SH NAME
bdftruncate \- generate truncated BDF font from ISO 10646-1-encoded BDF font
.SH SYNOPSIS
.B bdftruncate
.I threshold
.B <
.I source.bdf
.B >
.I destination.bdf
.sp
.B bdftruncate
[
.B \-\-help
|
.B \-\-version
]
.SH DESCRIPTION
.B bdftruncate
allows one to generate from an ISO10646-1 encoded BDF font other ISO10646-1
BDF fonts in which all characters above a threshold code value are stored
unencoded.  This is often desirable because the Xlib API and X11 protocol
data structures used for representing font metric information are extremely
inefficient when handling sparsely populated fonts.
.SH OPTIONS
.TP 12
.B \-\-help
prints usage message and exits.
.TP 12
.B \-\-version
prints version information and exits.
.SH EXAMPLE
The command
.RS
bdftruncate 0x3200 < 6x13.bdf > 6x13t.bdf
.RE
will generate the file 6x13t.bdf in which all glyphs with codes
>= 0x3200 will only be stored unencoded (i.e., they are encoded at
codepoint -1).
.SH "SEE ALSO"
.BR ucs2any (1)
.SH AUTHOR
.B bdftruncate
was written by Markus Kuhn.
.PP
Branden Robinson wrote this manual page, originally for the Debian Project.


xserver/ChangeLog
commit e597f0119cd69b6d9edf86d06d941468f90d8e6d
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 27 21:07:17 2012 -0800

    configure.ac: 1.11.4
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit 6c491463ee7a3637c8aae7ddae9625b36b792078
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 20 23:54:58 2012 -0800

    1.11.3.902 (1.11.4 RC2)
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit b747a91485184649c05aba6fa10734f70e926032
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Wed Jan 18 11:47:21 2012 -0800

    XQuartz: Bump bundle version to 2.7.1
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit a1ed67fa0551ee2b39d002d3b786a2017b37c7ac
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Fri Jan 6 13:20:45 2012 +1000

    render: don't bother with animated cursors on floating slaves (#39989)
    
    X.Org Bug 39989 <http://bugs.freedesktop.org/show_bug.cgi?id=39989>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit bbb6b8c834e0e1491ca14403b5d0840dd14380d3)

commit f89c39e6c9e20ffc3c1ae228ee601c2ab9007ff4
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Mon Jan 9 15:04:10 2012 +1000

    dix: Update pointer limits for floating devices too (#43635)
    
    When the screen is restructured, the pointer limits need to be reset for
    floating slave devices as well, not just for master pointers. Only skip
    devices that don't have a cursor (attached slaves and keyboard)
    
    Bug reproducer: float an absolute slave device, rotate the screen - the
    device is now confined to a section of the screen only.
    
    X.Org Bug 43635 <http://bugs.freedesktop.org/show_bug.cgi?id=43635>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Chase Douglas <chase.douglas@canonical.com>
    (cherry picked from commit 35bd77e9d0701daae87d681900d749604fc6471f)

commit dea9377a257a402b2c7fe4a96c958d951a7d4579
Author: Zhigang Gong <zhigang.gong@linux.intel.com>
Date:   Wed Jan 4 07:01:19 2012 +0000

    mi/mibitblt: Fix an overflow bug of bit shift.
    
    When depth equal to 32 and planeMask equal to 0, the overflow will
    occur and cause the pixmap can't be cleared. There are some test
    cases in XTS hit this bug, and this fix can eliminate the corresponding
    failures.
    
    Signed-off-by: Zhigang Gong <zhigang.gong@linux.intel.com>
    Reviewed-by: Adam Jackson <ajax@redhat.com>
    Signed-off-by: Keith Packard <keithp@keithp.com>
    (cherry picked from commit f4956faab9ccf9aba6cf9603f4489f5dad19a347)

commit d2db5100b1bafeeea8c1febd390631c8f8abb094
Author: Alan Coopersmith <alan.coopersmith@oracle.com>
Date:   Fri Oct 28 21:29:50 2011 -0700

    Convert DetermineClientCmd to use strdup instead of malloc+strncpy
    
    *cmdname is initialized to NULL earlier in the function, so it's
    okay to overwrite it with NULL if strdup fails, don't need that
    extra check.
    
    Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit 780133f9ae7fada462714b47e79d26075bbd9abe)

commit 5b28bcb340e279e7f4e76617e86b439092367690
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 21:23:12 2011 -0500

    dix: Tune dixLookupDrawable for success
    
    The vast vast vast majority of resource lookups are successful.  Move some
    work to the error paths so we don't punish success.
    
    Before:
    40000000 trep @   0.0009 msec (1109091.3/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2072652.2/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1148346.9/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2091666.1/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 8f9bdfd293ad8e45755efe8d764b4dcc2a724f51)

commit e1ea93fec93b13d18df4451210aaaf322b769b6a
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 20:23:40 2011 -0500

    dix: Tune dtrace hooks around Dispatch
    
    Don't call LookupMajorName if the hooks aren't active, it's quite expensive.
    
    Before:
    40000000 trep @   0.0009 msec (1087458.5/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2012238.6/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1109091.3/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2072652.2/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 83a98543b58c661a22574a6f8d8f9d777c0955b8)

commit 9eaa0b5c89f645c50ea90e07e8b49ebacf4349a0
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 21:00:05 2011 -0500

    xace: ricer tuning for XaceHook
    
    gcc doesn't want to hoist the check for XaceHooks[hook] != NULL above the
    varargs code for some reason, so do it ourselves.
    
    Before:
    40000000 trep @   0.0010 msec (1050420.2/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (1921147.6/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1087458.5/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2012238.6/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 11977fab546da35d579ebe285e3c26864007805e)

commit a3b6ecc893e17028cfbe16d8d7981c9037fc0766
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Tue Jan 3 10:45:45 2012 +1000

    xfree86: split warning about missing identifier or input driver
    
    Check for identifier first and bail if it's missing (also remove the current

font/util/bdftruncate.c
/*-
 * Copyright (c) 2006 Martin Husemann.
 * Copyright (c) 2007 Joerg Sonnenberger.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior
 *    written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * This program is derived (in a straight forward way) from
 * bdftruncate.pl -- Markus Kuhn <http://www.cl.cam.ac.uk/~mgk25/>
 *
 * This utility allows you to generate from an ISO10646-1 encoded
 * BDF font other ISO10646-1 BDF fonts in which all characters above
 * a threshold code value are stored unencoded.
 */

#include "config.h"

#include <ctype.h>
#include <errno.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifndef __has_attribute
# define __has_attribute(x) 0  /* Compatibility with older compilers. */
#endif
#ifndef __has_c_attribute
# define __has_c_attribute(x) 0  /* Compatibility with pre-C23 compilers. */
#endif

#if __has_c_attribute(noreturn)
# define ATTR_NORETURN [[noreturn]]
#elif __has_attribute(noreturn) \
    || (defined(__GNUC__) && ((__GNUC__ * 100 + __GNUC_MINOR__) >= 205)) \
    || (defined(__SUNPRO_C) && (__SUNPRO_C >= 0x590))
# define ATTR_NORETURN __attribute((noreturn))
#else
# define ATTR_NORETURN
#endif

static int iswide(unsigned int);
ATTR_NORETURN static void usage(int);
static int parse_threshold(const char *str, unsigned long *threshold);

static int
parse_threshold(const char *str, unsigned long *threshold)
{
	int base;
	char *end_ptr;

	if (!isdigit((unsigned char)*str))
		return 1;
	if (str[0] == '0' && (str[1] == 'x' || str[1] == 'X'))
		base = 16;
	else
		base = 10;

	errno = 0;
	*threshold = strtoul(str, &end_ptr, base);
	if (errno != 0 || *threshold == 0)
		return 1;
	return 0;
}

static void
process_line(const char *line, int removewide, unsigned long threshold)
{
	if (strncmp(line, "ENCODING", 8) == 0) {
		unsigned long enc;
		const char *v;

		v = line + 9;

		while (*v && isspace((unsigned char)(*v)))
			++v;
		enc = strtoul(v, NULL, 10);
		/* XXX Check for line-ending? */
		if (enc >= threshold || (removewide && iswide(enc))) {
			printf("ENCODING -1\n");
		} else {
			fputs(line, stdout);
		}
		return;
	}
	if (strncmp(line, "STARTFONT", 9) == 0) {
		fputs(line, stdout);
		printf("COMMENT AUTOMATICALLY GENERATED FILE. DO NOT EDIT!\n"
		    "COMMENT In this version of the font file, "
		    "all characters >= U+%04lx are\n"
		    "COMMENT not encoded to keep XFontStruct small.\n",
		    threshold);
		return;
	}
	if (strncmp(line, "COMMENT", 7) == 0) {
		const char *v = line + 8;

		while (*v && isspace((unsigned char)(*v)))
			v++;
		if (strncmp(v, "$id: ", 5) == 0 ||
		    strncmp(v, "$Id: ", 5) == 0) {
		    	const char *id = strchr(v+1, '$');
		    	if (id) {
		    		printf("COMMENT Derived from %.*s",
				     (int)(id - v - 4), v + 5);
				return;
		    	}
		}
	}
	fputs(line, stdout);
}

int
main(int argc, char **argv)
{
	int removewide;
	unsigned long threshold;
	int opt_minus_w = 0;
	int opt_plus_w = 0;
	char *line, *input_ptr;
	size_t line_len, rest_len;

	--argc;
	++argv;
	if (argc == 0)
		usage(EXIT_FAILURE);

	if (strcmp(*argv, "-w") == 0 || strcmp(*argv, "+w") == 0) {
		if (**argv == '-')
			opt_minus_w = 1;
		else
			opt_plus_w = 1;
		--argc;
		++argv;
	}
	else if (strcmp(*argv, "--help") == 0) {
		usage(EXIT_SUCCESS);
	}
	else if (strcmp(*argv, "--version") == 0) {
		puts(PACKAGE_STRING);
		exit(EXIT_SUCCESS);
	}

	if (argc != 1 || (opt_plus_w && opt_minus_w))
		usage(EXIT_FAILURE);
	if (parse_threshold(*argv, &threshold)) {
		fprintf(stderr, "Illegal threshold %s\n", *argv);
		usage(EXIT_FAILURE);
	}

	if (opt_minus_w)
		removewide = 1;
	else if (opt_plus_w)
		removewide = 0;
	else
		removewide = (threshold <= 0x3200);

	line_len = 1024;
	if ((line = malloc(line_len)) == NULL) {
		fprintf(stderr, "malloc failed");
		exit(EXIT_FAILURE);
	}

	for (;;) {
		if (fgets(line, line_len, stdin) == NULL)
		     break;
		while (strlen(line) == line_len - 1 && !feof(stdin)) {
			if (line_len > SSIZE_MAX) {
				fprintf(stderr, "input line too large");
				exit(EXIT_FAILURE);
			}
			line = realloc(line, line_len * 2);
			if (line == NULL) {
				fprintf(stderr, "realloc failed");
				exit(EXIT_FAILURE);
			}
			input_ptr = line + line_len - 1;
			rest_len = line_len + 1;
			line_len *= 2;
			if (fgets(input_ptr, rest_len, stdin) == NULL) {
				/* Should not happen, but handle as EOF */
				break;
			}
		}
		process_line(line, removewide, threshold);
	}

	return EXIT_SUCCESS;
}

/*
 * Subroutine to identify whether the ISO 10646/Unicode character code
 * ucs belongs into the East Asian Wide (W) or East Asian FullWidth
 * (F) category as defined in Unicode Technical Report #11.
 */
static int
iswide(unsigned int ucs)
{
    return (ucs >= 0x1100 &&
            (ucs <= 0x115f ||                   /* Hangul Jamo */
             (ucs >= 0x2e80 && ucs <= 0xa4cf &&
              (ucs & ~0x0011) != 0x300a && ucs != 0x303f) || /* CJK .. Yi */
             (ucs >= 0xac00 && ucs <= 0xd7a3) || /* Hangul Syllables */
             (ucs >= 0xf900 && ucs <= 0xfaff) || /* CJK Comp. Ideographs */
             (ucs >= 0xfe30 && ucs <= 0xfe6f) || /* CJK Comp. Forms */
             (ucs >= 0xff00 && ucs <= 0xff5f) || /* Fullwidth Forms */
             (ucs >= 0xffe0 && ucs <= 0xffe6) ||
             (ucs >= 0x20000 && ucs <= 0x2ffff)));
}

static void
usage(int exitstatus)
{
	fprintf(stderr,
	    "Usage: bdftruncate [+w|-w] threshold <source.bdf >destination.bdf\n"
            "   or: bdftruncate [--help|--version]\n"
	    "\n"
	    "Example:\n"
	    "\n"
	    "  bdftruncate 0x3200 <6x13.bdf >6x13t.bdf\n"
	    "\n"
	    "will generate the file 6x13t.bdf in which all glyphs with codes\n"
	    ">= 0x3200 will only be stored unencoded (i.e., ENCODING -1).\n"
	    "Option -w removes East Asian Wide and East Asian FullWidth characters\n"
	    "(default if threshold <= 0x3200), and option +w keeps them.\n");
	exit(exitstatus);
}
